synth/src/sampler.rs

```diff
@@ -33,103 +33,132 @@ impl Sampler {
+}
+
+impl SampleStrategy {
+    fn new(collection_name: Option<Name>, target: usize) -> Self {
+        match collection_name {
+            None => SampleStrategy::Namespace(NamespaceSampleStrategy {
+                target
+            }),
+            Some(name) => SampleStrategy::Collection(CollectionSampleStrategy {
+                name,
+                target,
+            })
         }
+    }
 
-        let mut rng = rand::rngs::StdRng::seed_from_u64(seed);
-        let mut model = self.graph.try_aggregate();
+    fn sample<R: Rng>(&self, model: Aggregate<Graph>, rng: R) -> Result<Value> {
+        match self {
+            SampleStrategy::Namespace(nss) => nss.sample(model, rng),
+            SampleStrategy::Collection(css) => css.sample(model, rng)
+        }
+    }
+}
 
+struct NamespaceSampleStrategy {
+    target: usize,
+}
+
+impl NamespaceSampleStrategy {
+    fn sample<R: Rng>(&self, mut model: Aggregate<Graph>, mut rng: R) -> Result<Value> {
         let mut generated = 0;
         let mut out = BTreeMap::new();
+        let progress_bar = sampler_progress_bar(self.target as u64);
 
-        let progress_bar = Self::sampler_progress_bar(target as u64);
-
-        while generated < target {
-            let start_of_round = generated;
-            let serializable = OwnedSerializable::new(model.try_next_yielded(&mut rng)?);
-            let mut value = match serde_json::to_value(&serializable)? {
-                Value::Object(map) => map,
-                _ => {
-                    return Err(
-                        failed!(target: Release, Unspecified => "generated data is not a namespace"),
-                    )
-                }
-            };
-
-            if let Some(name) = collection_name.as_ref() {
-                let collection_value = value.remove(name.as_ref()).ok_or_else(|| {
-                    failed!(
-                        target: Release,
-                        "generated namespace does not have a collection '{}'",
-                        name
-                    )
-                })?;
-                let vec = value_as_array(name.as_ref(), collection_value)?;
-                generated += vec.len();
-                match out.entry(name.to_string()) {
-                    Entry::Vacant(e) => {
-                        e.insert(vec);
-                    }
-                    Entry::Occupied(mut o) => {
-                        o.get_mut().extend(vec);
-                    }
-                }
-            } else {
-                value.into_iter().try_for_each(|(collection, value)| {
-                    value_as_array(&collection, value).map(|vec| {
-                        generated += vec.len();
-                        match out.entry(collection) {
-                            Entry::Vacant(e) => {
-                                e.insert(vec);
+        while generated < self.target {
+            let next = model.complete(&mut rng)?;
+            as_object(next)?
```

Comment:
This stuff could really do with some comments why the implementation was done this way.